db/event_ref.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package db
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"firebase.google.com/go/internal"
+)
+
+// Listen ...
+func (r *Ref) Listen(ctx context.Context) (*SnapshotIterator, error) {
+
+	sseDataChan := make(chan string) // server-sent event data channel
+
+	var opts []internal.HTTPOption
+	opts = append(opts, internal.WithHeader("Cache-Control", "no-cache"))
+	opts = append(opts, internal.WithHeader("Accept", "text/event-stream"))
+	opts = append(opts, internal.WithHeader("Connection", "keep-alive"))
+
+	resp, err := r.sendListen(ctx, "GET", opts...)
+
+	if err != nil {
+		return &SnapshotIterator{active: false}, err
+	}
+
+	snapshot, err := getInitialNodeSnapshot(resp)
+
+	if err != nil {
+		return &SnapshotIterator{active: false}, err
+	}
+
+	go r.startListeningWithReconnect(ctx, opts, resp, sseDataChan)
+
+	return &SnapshotIterator{
+		Snapshot:    snapshot,
+		SSEDataChan: sseDataChan,
+		resp:        resp, //*http.Response
+		active:      true,
+	}, err
+
+} // Listen()
+
+// return initial snapshot (JSON-encoded string) from Ref.Path node location
+func getInitialNodeSnapshot(resp *http.Response) (string, error) {
+
+	var b []byte
+
+	scanner := bufio.NewScanner(resp.Body)
+
+	if scanner.Scan() == true {
+
+		b = scanner.Bytes()
+
+		if "event: put" == string(b) {
+
+			if scanner.Scan() == true {
+				b = scanner.Bytes()
+				s := string(b)
+
+				var snapshotMap map[string]interface{}
+
+				err := json.Unmarshal([]byte(s[6:]), &snapshotMap)
```

Comment:
Better to comment why you've used a constant here

Reply:
Thanks. I've added comments from a recent commit.

Reply:
@thatreguy I think you need to sign a CLA https://cla.developers.google.com/